# Aula 3

**Susana Bouchardet [susana.bouchardet@databootcamp.com.br]**
![panda](https://www.writeraccess.com/wp-content/uploads/2014/11/blog-happy-panda.jpg)
______


 ## 0 - Revisão

In [1]:
# carregando os dados
import pandas
titanic = pandas.read_csv("titanic.csv")

In [2]:
# selecionando subconjunto de dados
titanic.loc[10:20,["Name","Pclass"]]

Name  Pclass
10                    Sandstrom, Miss. Marguerite Rut       3
11                           Bonnell, Miss. Elizabeth       1
12                     Saundercock, Mr. William Henry       3
13                        Andersson, Mr. Anders Johan       3
14               Vestrom, Miss. Hulda Amanda Adolfina       3
15                   Hewlett, Mrs. (Mary D Kingcome)        2
16                               Rice, Master. Eugene       3
17                       Williams, Mr. Charles Eugene       2
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...       3
19                            Masselmani, Mrs. Fatima       3
20                               Fynney, Mr. Joseph J       2

In [3]:
# Ordenando uma série
idades = titanic.Age[6:16]
# usando sorted [Perdemos os índices da série]
sorted(idades)

[2.0, 4.0, 14.0, 14.0, 20.0, 27.0, 39.0, 54.0, 55.0, 58.0]

In [4]:
# usando sort_values() [mantemos os índices da série] \o/
idades.sort_values()

7      2.0
10     4.0
9     14.0
14    14.0
12    20.0
8     27.0
13    39.0
6     54.0
15    55.0
11    58.0
Name: Age, dtype: float64

In [5]:
# ordenação decrescente
sorted(idades, reverse=True)

[58.0, 55.0, 54.0, 39.0, 27.0, 20.0, 14.0, 14.0, 4.0, 2.0]

In [6]:
idades.sort_values(ascending=False)

11    58.0
15    55.0
6     54.0
13    39.0
8     27.0
12    20.0
14    14.0
9     14.0
10     4.0
7      2.0
Name: Age, dtype: float64

In [7]:
# Apagando colunas
# usando del
titanic1 = titanic.copy()
del titanic1["Age"]
titanic1.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [8]:
# usando drop
titanic2 = titanic.drop(["Age"],axis=1)
titanic2.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [9]:
# o drop também apaga linas
titanic3 = titanic.drop([1,2,3,4], axis=0)
titanic3.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
5      0     330877   8.4583   NaN        Q  
6      0      17463  51.8625   E46        S  
7      1     349909  21.0750   NaN        S  
8      2     347742  11.1333   NaN        S

In [10]:
# Tratando valores nulos
# OBS.: lembrando como contamos os nulos: idades.isnull().sum()

idades = titanic.Age
print("""ANTES:
  Quantidade de nulos {};
  Tamanho da Série {}""".format(idades.isnull().sum(), idades.size))

# 1. Apagando
idades_sem_nulos = idades.dropna()
print("""DEPOIS:
  Quantidadede nulos {};
  Tamanho da Série {}""".format(idades_sem_nulos.isnull().sum(),idades_sem_nulos.size ))


ANTES:
  Quantidade de nulos 177;
  Tamanho da Série 891
DEPOIS:
  Quantidadede nulos 0;
  Tamanho da Série 714


In [11]:
# 2. Preenchendo
idades_nulos_preenchidos = idades.fillna(-1)
print("""DEPOIS:
  Quantidadede nulos {};
  Tamanho da Série {}""".format(idades_nulos_preenchidos.isnull().sum(),idades_nulos_preenchidos.size ))


DEPOIS:
  Quantidadede nulos 0;
  Tamanho da Série 891


In [12]:
# Agrupar dados com valores iguais
agrupados_por_sexo = titanic.groupby("Sex")["PassengerId"].count()
agrupados_por_sexo

Sex
female    314
male      577
Name: PassengerId, dtype: int64

In [13]:
# Agrupando usando apply
def funcao_agrupamento(valores):
    return valores.Sex.size

agrupados_por_sexo = titanic.groupby("Sex").apply(funcao_agrupamento)
agrupados_por_sexo

Sex
female    314
male      577
dtype: int64

In [14]:
# Alguns agrupamentos de valores categóricos (Categórical)
categ = pandas.Categorical(titanic["Survived"])
categ.describe()

counts     freqs
categories                  
0              549  0.616162
1              342  0.383838

In [15]:
type(categ)

pandas.core.categorical.Categorical

In [16]:
# usando o categorical
categ_renomeado = categ.rename_categories(["Morreu","Sobreviveu"])
categ_renomeado.describe()

counts     freqs
categories                  
Morreu         549  0.616162
Sobreviveu     342  0.383838

## 1 - Tranformar contínuos em categóricos

Já fizemos isso com o apply, __Mas existe um modo mais simples__

usando o `cut`

vamos transformar os dados de Idade em categóricos novamente, mas usando o `cut`

In [17]:

ranges = [titanic.Age.min()-1,10,18,40,titanic.Age.max()]
label = ["criança","adolescente","adulto","idoso"]

serie_faixa_etaria = pandas.cut(titanic.Age,ranges,labels=label)
serie_faixa_etaria.head()

0    adulto
1    adulto
2    adulto
3    adulto
4    adulto
Name: Age, dtype: category
Categories (4, object): [criança < adolescente < adulto < idoso]

O range são os valores que delimitam os intervalos.
Nesse caso as regras são
- criança:<br>
    $idade_{minima}-1 < idade \leqslant 10$
    
    
- adolescente:<br>
    $10 < idade \leqslant 18$
    
    
- adulto:<br>
    $18 < idade \leqslant 40$
    
    
- idoso:<br>
    $40 < idade \leqslant idade_{maxima}$


Vamos agora fazer a análise categórica desses valores.

In [18]:
# Criando a coluna FaixaDeIdade
titanic["FaixaDeIdade"] = serie_faixa_etaria

# Criando a analise categórica
categ = pandas.Categorical(titanic["FaixaDeIdade"])
categ.describe()

counts     freqs
categories                   
criança          64  0.071829
adolescente      75  0.084175
adulto          425  0.476992
idoso           150  0.168350
NaN             177  0.198653

Se não quisermos explicitar os intervalos, podemos dividir em $n$ grupos com intervalor iguals

In [19]:
pandas.cut(titanic.Age,bins=4).head(10)

0    (20.315, 40.21]
1    (20.315, 40.21]
2    (20.315, 40.21]
3    (20.315, 40.21]
4    (20.315, 40.21]
5                NaN
6    (40.21, 60.105]
7     (0.34, 20.315]
8    (20.315, 40.21]
9     (0.34, 20.315]
Name: Age, dtype: category
Categories (4, interval[float64]): [(0.34, 20.315] < (20.315, 40.21] < (40.21, 60.105] < (60.105, 80.0]]

<span style="color:blue">__Vamos transformar Fare em um segmento também?__ </span>

In [20]:
result = pandas.cut(titanic.Fare, bins = 5)
cat = pandas.Categorical(result)
cat.describe()

counts     freqs
categories                          
(-0.512, 102.466]      838  0.940516
(102.466, 204.932]      33  0.037037
(204.932, 307.398]      17  0.019080
(307.398, 409.863]       0  0.000000
(409.863, 512.329]       3  0.003367

## 2 - Transformando Séries do tipo _String_

Podemos tratar as Strings fazendo um _apply_...

__mas existe uma maneira mais simples para aplicar algumas tranformações!__

Vamos começar com uma tranformação simples: retirar as vírgulas

### 2.1 - _replace_ na série

In [21]:
# Objetivo 
#  Braund, Mr. Owen Harris => Braund Mr. Owen Harris
nomes_sem_virgula = titanic.Name.str.replace(",","")
nomes_sem_virgula.head()

0                               Braund Mr. Owen Harris
1    Cumings Mrs. John Bradley (Florence Briggs Tha...
2                                Heikkinen Miss. Laina
3          Futrelle Mrs. Jacques Heath (Lily May Peel)
4                              Allen Mr. William Henry
Name: Name, dtype: object

### 2.1 - _split_ na série
Agora vamos deixar apenas o sobrenome

In [22]:
# Split por , e tranformar o resultado em um Dataframe
serie_nome_dividido = titanic.Name.str.split(",")
type(serie_nome_dividido)

pandas.core.series.Series

In [23]:
serie_nome_dividido.head()

0                           [Braund,  Mr. Owen Harris]
1    [Cumings,  Mrs. John Bradley (Florence Briggs ...
2                            [Heikkinen,  Miss. Laina]
3      [Futrelle,  Mrs. Jacques Heath (Lily May Peel)]
4                          [Allen,  Mr. William Henry]
Name: Name, dtype: object

Ficamos com uma __série de listas__.
Como pegamos o primeiro valor dessa lista agora?

Só conseguimos fazer isso com __apply__.

Uma Série de listas com tamanhos iguais faz sentido ser interpretada como um dataframe!

In [24]:
dataframe_nome_dividido = titanic.Name.str.split(",", expand=True)
type(dataframe_nome_dividido)

pandas.core.frame.DataFrame

In [25]:
dataframe_nome_dividido.head()

0                                            1
0     Braund                              Mr. Owen Harris
1    Cumings   Mrs. John Bradley (Florence Briggs Thayer)
2  Heikkinen                                  Miss. Laina
3   Futrelle           Mrs. Jacques Heath (Lily May Peel)
4      Allen                            Mr. William Henry

In [26]:
dataframe_nome_dividido[0].head()

0       Braund
1      Cumings
2    Heikkinen
3     Futrelle
4        Allen
Name: 0, dtype: object

Com o resultado, agora podemos criar uma coluna _Sobrenome_

In [27]:
titanic["Sobrenome"] = dataframe_nome_dividido[0]

In [28]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked FaixaDeIdade  Sobrenome  
0      0         A/5 21171   7.2500   NaN        S       adulto     Braund  
1      0          PC 17599  71.2833   C85        C       adulto    Cumings  
2      0  STON/O2. 3101282   7.9250   NaN        S       adulto  Heikkinen  
3      0            113803  53.1000  C123        S       adulto   Futrelle  
4      0            373450   8.0500   NaN        S       adulto      Allen

__E se as listas da série não tiverem tamanhos iguais?__

A diferença se tranforma em valor nulo!

In [29]:
# A diferença se tranforma em valor nulo
dataframe_nome_dividido_diferente = titanic.Name.str.split("(", expand=True)
dataframe_nome_dividido_diferente.head()

0                        1     2
0        Braund, Mr. Owen Harris                     None  None
1    Cumings, Mrs. John Bradley   Florence Briggs Thayer)  None
2         Heikkinen, Miss. Laina                     None  None
3  Futrelle, Mrs. Jacques Heath            Lily May Peel)  None
4       Allen, Mr. William Henry                     None  None

<span style="color:blue">__Vamos agora criar uma coluna que contenha o nome das mulheres (Nome entre parênteses)__</span>

In [30]:
# Split por "(" para pegar o nome da mulher
dataframe_nome_dividido_mulheres_1 = titanic.Name.str.split("(", expand=True)

# replace de ")" por "" para remover o final do parentese no campo
dataframe_nome_dividido_mulheres_2 = dataframe_nome_dividido_mulheres_1[1].str.replace(")","")

dataframe_nome_dividido_mulheres_2.head()

0                      None
1    Florence Briggs Thayer
2                      None
3             Lily May Peel
4                      None
Name: 1, dtype: object

In [31]:
titanic["NomeMulher"] = dataframe_nome_dividido_mulheres_2
titanic[["Name","NomeMulher"]].head(10)

Name  \
0                            Braund, Mr. Owen Harris   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2                             Heikkinen, Miss. Laina   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4                           Allen, Mr. William Henry   
5                                   Moran, Mr. James   
6                            McCarthy, Mr. Timothy J   
7                     Palsson, Master. Gosta Leonard   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9                Nasser, Mrs. Nicholas (Adele Achem)   

                  NomeMulher  
0                       None  
1     Florence Briggs Thayer  
2                       None  
3              Lily May Peel  
4                       None  
5                       None  
6                       None  
7                       None  
8  Elisabeth Vilhelmina Berg  
9                Adele Achem

### Tratando os nulos na funcao do `apply`

<span style="color:blue">__Nós temos nome das mulheres na coluna `NomeMulher`. Podemos então substituir na coluna `Name` o nome das mulheres pelo nome na coluna `NomeMulher`.__</span>

Nesse caso não teremos problema de perder informação, pois o sobrenome esta salvo, então podemos identificar a família da mulher.

Dica:
- O `pandas` tem um teste para valores nulos: `pandas.isnull(valor)`. Ele retorna verdadeiro se for nulo e falso se não for.
- A função que será aplicada deverá testar os valores de entrada para conteudo nulo com um `if`

In [32]:
def substitui_nome(linha_do_dataframe):
    nome_mulher = linha_do_dataframe["NomeMulher"]
    nome = linha_do_dataframe["Name"]
    if not pandas.isnull(nome_mulher):
        return nome_mulher
    else:
        return nome
    
nova_coluna_name = titanic.apply(substitui_nome, axis=1)
nova_coluna_name.head()

0     Braund, Mr. Owen Harris
1      Florence Briggs Thayer
2      Heikkinen, Miss. Laina
3               Lily May Peel
4    Allen, Mr. William Henry
dtype: object

In [33]:
titanic["Name"]=nova_coluna_name
titanic.head()

PassengerId  Survived  Pclass                      Name     Sex   Age  \
0            1         0       3   Braund, Mr. Owen Harris    male  22.0   
1            2         1       1    Florence Briggs Thayer  female  38.0   
2            3         1       3    Heikkinen, Miss. Laina  female  26.0   
3            4         1       1             Lily May Peel  female  35.0   
4            5         0       3  Allen, Mr. William Henry    male  35.0   

   SibSp  Parch            Ticket     Fare Cabin Embarked FaixaDeIdade  \
0      1      0         A/5 21171   7.2500   NaN        S       adulto   
1      1      0          PC 17599  71.2833   C85        C       adulto   
2      0      0  STON/O2. 3101282   7.9250   NaN        S       adulto   
3      1      0            113803  53.1000  C123        S       adulto   
4      0      0            373450   8.0500   NaN        S       adulto   

   Sobrenome              NomeMulher  
0     Braund                    None  
1    Cumings  Florence Briggs Thayer  
2  Heikkinen                    None  
3   Futrelle           Lily May Peel  
4      Allen                    None

__Sabendo disso, conseguimos resolver juntos o exercício final da Aula 2! __

<span style="color:blue">__Substituir as idades nulas pela média das idades do sexo__</span>

Ex.: Se a média da idade das mulheres é 25, toda mulher com idade nula deve ser preenchida com 25

In [34]:
media_mulher = titanic[titanic.Sex == "female"].Age.mean()
media_homen = titanic[titanic.Sex == "male"].Age.mean()

print("""
Media de idade das mulheres: {}
Media de idade dos homens: {}""".format(media_mulher,media_homen))


Media de idade das mulheres: 27.915708812260537
Media de idade dos homens: 30.72664459161148


In [35]:
def preenche_idade(linha):
    idade = linha["Age"]
    sexo = linha["Sex"]
    
    if pandas.isnull(idade):
        
        if sexo == "femane":
            return media_mulher
        else:
            return media_homen
        
    else:
        return idade
    

idades_preenchidas = titanic.apply(preenche_idade,axis=1)

print("""
Quantidade de nulos após o preenchimento: {}
""".format(idades_preenchidas.isnull().sum()))       


Quantidade de nulos após o preenchimento: 0



In [36]:
# Vamos substituir a idade atual pelo resultado do apply
titanic["Age"] = idades_preenchidas

## 4 - Tabela de frequência

Para entender a correlação entre duas variaveis precisamos coloca-las lado a lado.

O Pandas nos da uma funcionalidade de crosstab(tabela de frequência).


__Vamos fazer a tabela de frequência entre os sobrenomes e as cabine, para validarmos se cada cabine possuia apenas uma família__

In [37]:
tabela_frequencia_cabine_sobrenome = pandas.crosstab(titanic.Cabin, titanic.Sobrenome)
tabela_frequencia_cabine_sobrenome.head()

Sobrenome  Allen  Allison  Anderson  Andrews  Appleton  Astor  Aubart  Ball  \
Cabin                                                                         
A10            0        0         0        0         0      0       0     0   
A14            0        0         0        0         0      0       0     0   
A16            0        0         0        0         0      0       0     0   
A19            0        0         0        0         0      0       0     0   
A20            0        0         0        0         0      0       0     0   

Sobrenome  Barkworth  Baxter  ...    Van der hoef  Walker  Warren  Webber  \
Cabin                         ...                                           
A10                0       0  ...               0       0       0       0   
A14                0       0  ...               0       0       0       0   
A16                0       0  ...               0       0       0       0   
A19                0       0  ...               0       0       0       0   
A20                0       0  ...               0       0       0       0   

Sobrenome  White  Wick  Widener  Williams-Lambert  Woolner  Young  
Cabin                                                              
A10            0     0        0                 0        0      0  
A14            0     0        0                 0        0      0  
A16            0     0        0                 0        0      0  
A19            0     0        0                 0        0      0  
A20            0     0        0                 0        0      0  

[5 rows x 158 columns]

In [38]:
def cabin_with_family(line):
    soma = 0
    for i in line:
        if i != 0:
            soma = soma + 1
    return soma

familias_diferentes_por_cabine = tabela_frequencia_cabine_sobrenome.apply(cabin_with_family,axis=1)
familias_diferentes_por_cabine.sort_values(ascending=False).head()

Cabin
F33     3
D       3
E101    3
B5      2
D20     2
dtype: int64

## 3 - Salvando o Dataframe

In [39]:
# Preparando o Dataframe para ser salvo

# Como ja passamos para Name, vamos apagar a coluna NomeMulher diretamente do dataframe
del titanic["NomeMulher"]

In [40]:
# salvando
titanic.to_csv("titanic_novo.csv")

o dataframe pode ser salvo em outros formatos.

In [41]:
# Salvando em Json
titanic.to_json("titanic_novo.json")

_________
Vamos aprender a trabalhar com outro tipos de dados que não temos no dataset _Titanic_.

Para continuar então vamos criar nosso dataset!

<font size="5">http://bit.ly/2m7CqWf</font>

In [42]:
alunos = pandas.read_csv("http://bit.ly/2zrHNVW")

<font size="6">http://bit.ly/2zrHNVW</font>

## 4 - Tratando Datas

In [43]:
import datetime

alunos.dtypes

Nome       object
Data       object
sexo       object
altura    float64
dtype: object

Preenchemos um dado como data, mas quando lemos o csv ele foi interpretado como `object`.

Vamos convertes o dado para o tipo correto

In [44]:
novas_datas = alunos.Data.apply(pandas.to_datetime)
novas_datas

0    1994-04-27
1    1993-10-11
2    1990-11-08
3    1945-04-14
4    1986-11-09
5    1976-06-19
6    1987-08-31
7    1981-07-02
8    1988-07-11
9    1967-06-08
10   1956-01-08
11   1989-12-12
Name: Data, dtype: datetime64[ns]

In [45]:
pandas.to_datetime?

In [46]:
alunos.Data = novas_datas

In [47]:
alunos.dtypes

Nome              object
Data      datetime64[ns]
sexo              object
altura           float64
dtype: object

Agora o tipo da coluna é __datetime__, podemos então aplicar os métodos de __datetime__ na coluna!

In [48]:
# Série apenas com os meses
alunos.Data.dt.month

0      4
1     10
2     11
3      4
4     11
5      6
6      8
7      7
8      7
9      6
10     1
11    12
Name: Data, dtype: int64

In [49]:
# Série apenas com o ano
alunos.Data.dt.year

0     1994
1     1993
2     1990
3     1945
4     1986
5     1976
6     1987
7     1981
8     1988
9     1967
10    1956
11    1989
Name: Data, dtype: int64

In [50]:
# Série com o dia da semana 
# Ordem:
# 'Segunda-feira', 'Terça-feira', 'Quarta-feira', 'Quinta-feira', 'Sexta-feira', 'Sábado', 'Domingo'
alunos.Data.dt.weekday

0     2
1     0
2     3
3     5
4     6
5     5
6     0
7     3
8     0
9     3
10    6
11    1
Name: Data, dtype: int64

Podemos também, caso seja necessário, criar uma série com o tipo string com um formato especéifico na data

In [51]:
alunos.Data.dt.strftime("%d/%m/%Y")

0     27/04/1994
1     11/10/1993
2     08/11/1990
3     14/04/1945
4     09/11/1986
5     19/06/1976
6     31/08/1987
7     02/07/1981
8     11/07/1988
9     08/06/1967
10    08/01/1956
11    12/12/1989
Name: Data, dtype: object

Formatação de data no python segue um padrão de varias linguagens de programação.
Basta você montar a string de formatação substituindo os valores pelos simbolos correnpondentes.

Você pode [ver os simbolos aqui](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior).


__<span style="color:blue">Com as datas de nascimento no tipo certo, podemos calcular as idades das pessoas em nosso dataset!</span>__

Dica:
- Para acessar o ano do datetime: date.year

In [52]:
import datetime
today = datetime.datetime.today()

In [53]:
today.year

2017

In [54]:
def idade(linha):
    data = linha["Data"]
    return datetime.datetime.now().year - data.year

alunos.apply(idade, axis=1)

0     23
1     24
2     27
3     72
4     31
5     41
6     30
7     36
8     29
9     50
10    61
11    28
dtype: int64

## 5 - Unindo Datasets

Em muitos casos os dados não estarão todos no mesmo dataset. Para resolver isso podemos criar um dataset ja junção de outros. Para juntar dois datasets basta que tenhamos um dado em comum entre os dois.


Vamos fazer um dataframe com os dias da semana, e agregar essa informação ao dataframe alunos

In [55]:
# Criando um dataframe
dias_da_semana = [{"dia_da_semana":"Segunda", "numero":0},
                 {"dia_da_semana":"Terça", "numero":1},
                 {"dia_da_semana":"Quarta", "numero":2},
                 {"dia_da_semana":"Quinta", "numero":3},
                 {"dia_da_semana":"Sexta", "numero":4},
                 {"dia_da_semana":"Sábado", "numero":5},
                 {"dia_da_semana":"Domingo", "numero":6}]

dataframe_dias_da_semana = pandas.DataFrame(dias_da_semana)

dataframe_dias_da_semana

dia_da_semana  numero
0       Segunda       0
1         Terça       1
2        Quarta       2
3        Quinta       3
4         Sexta       4
5        Sábado       5
6       Domingo       6

In [56]:
# Criando a coluna com o dia da semana (número) no dataframe alunos

alunos["dia_da_semana"] = alunos.Data.dt.weekday
alunos

Nome       Data sexo  altura  dia_da_semana
0          Susana 1994-04-27    f    1.77              2
1          Daniel 1993-10-11    f    1.82              0
2            Laís 1990-11-08    f    1.68              3
3        Mauricio 1945-04-14    f    1.72              5
4       Alexandre 1986-11-09    m    1.75              6
5   Daniel Garcia 1976-06-19    m    1.69              5
6   Danila Moreno 1987-08-31    F    1.55              0
7          Felipe 1981-07-02    m    1.74              3
8         Adriano 1988-07-11    m    1.80              0
9             Cid 1967-06-08    m    1.75              3
10          Edson 1956-01-08    m    1.76              6
11   Gustavo Leão 1989-12-12    m    1.93              1

In [57]:
# Juntar os dataframes pelo campos dia_da_semana = numero
pandas.merge(alunos, dataframe_dias_da_semana, left_on="dia_da_semana", right_on="numero",how="inner")

Nome       Data sexo  altura  dia_da_semana_x dia_da_semana_y  \
0          Susana 1994-04-27    f    1.77                2          Quarta   
1          Daniel 1993-10-11    f    1.82                0         Segunda   
2   Danila Moreno 1987-08-31    F    1.55                0         Segunda   
3         Adriano 1988-07-11    m    1.80                0         Segunda   
4            Laís 1990-11-08    f    1.68                3          Quinta   
5          Felipe 1981-07-02    m    1.74                3          Quinta   
6             Cid 1967-06-08    m    1.75                3          Quinta   
7        Mauricio 1945-04-14    f    1.72                5          Sábado   
8   Daniel Garcia 1976-06-19    m    1.69                5          Sábado   
9       Alexandre 1986-11-09    m    1.75                6         Domingo   
10          Edson 1956-01-08    m    1.76                6         Domingo   
11   Gustavo Leão 1989-12-12    m    1.93                1           Terça   

    numero  
0        2  
1        0  
2        0  
3        0  
4        3  
5        3  
6        3  
7        5  
8        5  
9        6  
10       6  
11       1

![join](http://www.dofactory.com/Images/sql-joins.png)

<b style = "color:blue">
Exercício final:

1\. Criar uma dataframe que tenha uma coluna com categorias de altura e outra com a descrição.

| categorai | descricao |
|-----------|-----------|
| 1         | 1.5 - 1.6 |
| 2         | 1.6 - 1.7 |
| 3         | 1.7 - 1.8 |
| 4         | 1.8 - 1.9 |
| 5         | 1.9 - 2   |

2\. Categorizar as alturas dos alunos, assumindo o range de 1.5m até 2m.

| Nome   | altura | ... | categoria |
|--------|--------|-----|-----------|
| Susana | 1.77   | ... | 3         |


3\. Juntar as tabelas de maneira cada linha tenha as informações dos alunos e a descrição da categoria de altura.

| Nome   | altura | ... | categoria | descricao |
|--------|--------|-----|-----------|-----------|
| Susana | 1.77   | ... | 3         | 1.7 - 1.8 |
| null   | null   | ... | 1         | 1.5 - 1.6 |
| null   | null   | ... | 2         | 1.6 - 1.7 |
| null   | null   | ..  | 4         | 1.8 - 1.9 |
| null   | null   | ... | 5         | 1.9 - 2.0 |

</b>